src/device_trezor/device_trezor_base.cpp

```diff
@@ -0,0 +1,301 @@
+      device_locker.lock();
+      MTRACE("Device " << this->name << " LOCKed");
+    }
+
+    //lock the device for a long sequence
+    bool device_trezor_base::try_lock() {
+      MTRACE("Ask for LOCKING(try) for device " << this->name << " in thread ");
+      bool r = device_locker.try_lock();
+      if (r) {
+        MTRACE("Device " << this->name << " LOCKed(try)");
+      } else {
+        MDEBUG("Device " << this->name << " not LOCKed(try)");
+      }
+      return r;
+    }
+
+    //unlock the device
+    void device_trezor_base::unlock() {
+      MTRACE("Ask for UNLOCKING for device " << this->name << " in thread ");
+      device_locker.unlock();
+      MTRACE("Device " << this->name << " UNLOCKed");
+    }
+
+    /* ======================================================================= */
+    /*  Helpers                                                                */
+    /* ======================================================================= */
+
+    void device_trezor_base::require_connected(){
+      if (!m_transport){
+        throw exc::NotConnectedException();
+      }
+    }
+
+    void device_trezor_base::call_ping_unsafe(){
+      auto pingMsg = std::make_shared<messages::management::Ping>();
+      pingMsg->set_message("PING");
+
+      auto success = this->client_exchange<messages::common::Success>(pingMsg);  // messages::MessageType_Success
+      MDEBUG("Ping response " << success->message());
+      (void)success;
+    }
+
+    void device_trezor_base::test_ping(){
+      require_connected();
+
+      try {
+        this->call_ping_unsafe();
+
+      } catch(exc::TrezorException const& e){
+        MINFO("Trezor does not respond: " << e.what());
+        throw exc::DeviceNotResponsiveException(std::string("Trezor not responding: ") + e.what());
+      }
+    }
+
+    /* ======================================================================= */
+    /*                              TREZOR PROTOCOL                            */
+    /* ======================================================================= */
+
+    bool device_trezor_base::ping() {
+      AUTO_LOCK_CMD();
+      if (!m_transport){
+        MINFO("Ping failed, device not connected");
+        return false;
+      }
+
+      try {
+        this->call_ping_unsafe();
+        return true;
+
+      } catch(std::exception const& e) {
+        MERROR("Ping failed, exception thrown " << e.what());
+      } catch(...){
+        MERROR("Ping failed, general exception thrown" << boost::current_exception_diagnostic_information());
+      }
+
+      return false;
+    }
+
+    void device_trezor_base::on_button_request()
+    {
+      if (m_callback){
+        m_callback->on_button_request();
+      }
+    }
+
+    void device_trezor_base::on_pin_request(std::string & pin)
+    {
+      if (m_callback){
+        m_callback->on_pin_request(pin);
+      }
+    }
+
+    void device_trezor_base::on_passphrase_request(bool on_device, std::string & passphrase)
```

Comment:
You may want to use wipeable_string here, it looks like secret data. Maybe pin above too.